### PR TITLE
[WIP] Allow ordereddict usage for yaml mappings

### PR DESCRIPTION
**Pull request by bcoca**
##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

all
##### ANSIBLE VERSION

```
?.x
```

### CODE REVIEW

File: lib/ansible/plugins/inventory/yaml.py [hunk 1]
```diff
@@ -114,15 +114,15 @@ def _parse_group(self, group, group_data):
 
         self.inventory.add_group(group)
 
-        if isinstance(group_data, MutableMapping):
+        if isinstance(group_data, Mapping):
```
Comment (abadger) [inline lib/ansible/plugins/inventory/yaml.py]: This one is wrong.  (See line 123).

File: lib/ansible/plugins/inventory/yaml.py [hunk 2]
```diff
@@ -59,7 +59,7 @@
 '''
```
Comment (abadger) [inline lib/ansible/plugins/inventory/yaml.py:70]: MutableMapping => Mapping should be a separate PR.

**Comment by abadger** [comment]
Reviewed and gave feedback on slack:

> -1 to make this an end user-visible difference in behaviour when ordered dict is unavailable.  +1 to bundle the backport ordered dict module so behavior is the same everywhere (and I think there's at least one other place we can use it).  Also needs tests.

Additionally, If there's no drawback to using ordered dicts, I'd be tempted to take out the parameters allowing switching between the two and use ordereddicts always.

File: lib/ansible/module_utils/ordereddict.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+except ImportError:
+    from collections.abc import KeysView, ValuesView, ItemsView
+
+try:
+    from ansible.module_utils.six.moves._thread import get_ident as _get_ident
+except ImportError:
+    from ansible.module_utils.six.moves._dummy_thread import get_ident as _get_ident
+
+
+class OrderedDict(dict):
```
Comment (abadger) [inline lib/ansible/module_utils/ordereddict.py]: @sivel From the rest of the code, it looks like you probably have a reason for inheriting from dict instead of from MutableMapping.  Probably need a comment to explain that, though.

Comment (sivel) [reply lib/ansible/module_utils/ordereddict.py]: Honestly, it wasn't a choice that I specifically made.  This implementation came from https://github.com/ActiveState/code/tree/3b27230f418b714bc9a0f897cb8ea189c3515e99/recipes/Python/576693_Ordered_Dictionary_for_Py24

I only made some modifications to utilize some of our compat code.

File: lib/ansible/parsing/yaml/objects.py [hunk 4]
```diff
@@ -21,6 +21,17 @@
...
+try:
+    # included in standard lib from Python 2.7
+    from collections import OrderedDict as omap
+except ImportError:
+    try:
+        # in case we have the external lib
+        from orderedict import OrderedDict as omap
+    except ImportError:
+        # fallback to vendored version
+        from ansible.module_utils.orderedict import OrderedDict as omap
```
Comment (sivel) [inline lib/ansible/parsing/yaml/objects.py]: Just curious, why are we importing `as omap` here?

File: lib/ansible/parsing/yaml/objects.py [hunk 5]
```diff
@@ -21,6 +21,17 @@
 
 import yaml
 
+try:
+    # included in standard lib from Python 2.7
+    from collections import OrderedDict as omap
+except ImportError:
+    try:
+        # in case we have the external lib
+        from orderedict import OrderedDict as omap
```
Comment (sivel) [inline lib/ansible/parsing/yaml/objects.py]: Missing another `d`.  `ordereddict` instead of `orderedict`

File: lib/ansible/parsing/yaml/objects.py (hunk 4 above)
Comment (sivel) [inline lib/ansible/parsing/yaml/objects.py]: Missing another `d`. `ordereddict` instead of `orderedict`

File: lib/ansible/module_utils/ordereddict.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+        try:
+            self.__root
+        except AttributeError:
+            self.__root = root = []                     # sentinel node
+            root[:] = [root, root, None]
+            self.__map = {}
+        self.__update(*args, **kwds)
+
+    def __setitem__(self, key, value, dict_setitem=dict.__setitem__):
+        'od.__setitem__(i, y) <==> od[i]=y'
```
Comment (sivel) [inline lib/ansible/module_utils/ordereddict.py]: We may want to turn these single quoted docstrings into triple quoted, or just make them comments.

I imagine the false positive checks for `dict.iteritems` and such will disappear when that is done.

File: lib/ansible/module_utils/ordereddict.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+        Pairs are returned in LIFO order if last is true or FIFO order if false.
+
+        '''
+        if not self:
+            raise KeyError('dictionary is empty')
+        key = next(reversed(self) if last else iter(self))
+        value = self.pop(key)
+        return key, value
+
+    def __repr__(self, _repr_running={}):
```
Comment (sivel) [inline lib/ansible/module_utils/ordereddict.py]: Based on what this is doing, I assume that it is "unsafe" (using a dict as the default) purposefully.  It seems to be tracking some form of state.

File: lib/ansible/parsing/yaml/objects.py (hunk 4 above)
Comment (bcoca) [reply lib/ansible/parsing/yaml/objects.py]: mostly cause that is how yaml does it and i wanted to keep the same implementation

File: lib/ansible/parsing/yaml/objects.py (hunk 5 above)
Comment (bcoca) [reply lib/ansible/parsing/yaml/objects.py]: ???

Comment (bcoca) [reply lib/ansible/parsing/yaml/objects.py]: nvmd ... didn't see its a double D

**Comment by sivel** [comment]
I'm guessing the rest of the sanity issues just need explicit pylint disables, or added to ignore lists in the code smell tests.

**Comment by ghost** [comment]
Hi guys, do we have any quick workaround for this? facing similar issue

**Expected BODY**
```
            "body": {
                "analytics-cluster": [
                    {
                       "name": "Analytics-Cluster-1",
                        "connector-config": {
                            "ip-address": [
                                [
                                    "192.168.185.4", 
                                    "192.168.185.5", 
                                    "192.168.185.6", 
                                    "192.168.185.7"
                                ]
                            ]
                        }, 
                        "log-collector-config": {
                            "ip-address": [
                                [
                                    "192.168.185.20", 
                                    "192.168.185.21", 
                                    "192.168.185.22", 
                                    "192.168.185.23"
                                ]
                            ], 
                            "port": "1234"
                        }
                    }
                ]
            }, 

```
**ANSIBLE REORDERING  BODY - Moving name to end**
```
            "body": {
                "analytics-cluster": [
                    {
                        "connector-config": {
                            "ip-address": [
                                [
                                    "192.168.185.4", 
                                    "192.168.185.5", 
                                    "192.168.185.6", 
                                    "192.168.185.7"
                                ]
                            ]
                        }, 
                        "log-collector-config": {
                            "ip-address": [
                                [
                                    "192.168.185.20", 
                                    "192.168.185.21", 
                                    "192.168.185.22", 
                                    "192.168.185.23"
                                ]
                            ], 
                            "port": "1234"
                        }, 
                        "name": "Analytics-Cluster-1"
                    }
                ]
            }, 
```

File: lib/ansible/module_utils/ordereddict.py (hunk 3 above)
Comment (abadger) [reply lib/ansible/module_utils/ordereddict.py]: We probably want to inherit from MutableMapping instead of dict, then.

File: lib/ansible/module_utils/ordereddict.py (hunk 7 above)
Comment (abadger) [reply lib/ansible/module_utils/ordereddict.py]: It looks like we could do this with a class attribute (not an instance attribute) instead of a funny use of a parameter.

Comment (abadger) [reply lib/ansible/module_utils/ordereddict.py]:
``` python
class Test:
    a = {}
    def test(self, c={}):
        print(id(self.a))
        print(id(c))

a = Test()
b = Test()
a.test()
b.test()
a.test()
b.test()
```

Replacing ```c``` with ```a``` should have the same behaviour

File: lib/ansible/module_utils/ordereddict.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+        '''Initialize an ordered dictionary.  The signature is the same as
+        regular dictionaries, but keyword arguments are not recommended because
+        their insertion order is arbitrary.
+
+        '''
+        if not args:
+            raise TypeError("descriptor '__init__' of 'OrderedDict' object "
+                            "needs an argument")
+        self = args[0]
+        args = args[1:]
```
Comment (abadger) [inline lib/ansible/module_utils/ordereddict.py]: To solve the no-method-argument pylint warning, change the definition to
```
def __init__(self, *args, **kwds):
```
and remove lines 38-41 (the lines that sanity check and set self from args)

File: lib/ansible/module_utils/ordereddict.py (hunk 3 above)
Comment (bcoca) [reply lib/ansible/module_utils/ordereddict.py]: what would be the advantages/dis for either case?

File: lib/ansible/module_utils/ordereddict.py (hunk 8 above)
Comment (abadger) [reply lib/ansible/module_utils/ordereddict.py]: *remove lines 38-42.

File: lib/ansible/module_utils/ordereddict.py (hunk 3 above)
Comment (abadger) [reply lib/ansible/module_utils/ordereddict.py]: https://github.com/xonsh/xonsh/pull/125#discussion_r26907490

**Comment by abadger** [comment]
@bcoca, note... without ordereddict, we cannot merge this until mattclay has updated the test infrastructure for 2.7.  Might be better to restore the ordereddict bundling so we can add this sooner.

Also... I thought you put the ordereddict backport into module_utils so that modules could take advantage of it?  That would still be valid.

**Comment by BarbzYHOOL** [comment]
Why closed?

**Comment by AlanCoding** [comment]
@BarbzYHOOL This might not work on its own, or not work for all cases on its own.

As a note to myself, I intent to test in conjunction with https://github.com/ansible/ansible/pull/36644. I think that bcoca's PR here would have addressed the steps to reproduce given in https://github.com/ansible/ansible/issues/34861, but I see some other inventory file examples in this issue which are more complicated, some of which may need both.

**Comment by hnijbacker** [comment]
@bcoca I would like to know the status of this PR. This issue is closed, but the merge is not done.

**Comment by Kassec** [comment]
@bcoca same question here. This is a much needed feature for my team.

**Comment by splitice** [comment]
Status?
